lib/mastodon/plugin.rb

```diff
@@ -0,0 +1,112 @@
+# frozen_string_literal: true
+
+module Mastodon
+  class Plugin
+    NamedPath = Struct.new(:name, :path, :type)
+
+    attr_reader :actions, :paths, :assets
+
+    def initialize
+      @actions = Set.new
+      @paths   = Set.new
+      @assets  = Set.new
+    end
+
+    # defined by the plugin
+    def setup!
+      raise NotImplementedError.new, "Please define a setup! method"
```

Comment:
Couldn't assets and paths be defined using class methods instead of requiring them to be executed inside this method? E.g.:

```ruby
class MyPlugin < Mastodon::Plugin
  translation '/some/path'

  routes do
    get :my_extra_route, to: 'my_controller#index'
  end
end
```

Reply:
Sure, happy to do that.

Reply:
The methods as is don't actually run anything either, except to define a Proc which gets queued up in a list of 'stuff to do' when the plugin gets installed.

This means that we could introduce a concept of uninstalled or disabled plugins, which would still be evaluated but don't actually apply any changes to core unless they're marked as 'installed' - one thing I think would be good to include in an early iteration is a 'safe mode'
```
foreman start --safe
// or maybe
foreman start --disable-plugins example
foreman start --disable-plugins all
```
which would allow admins to more easily diagnose whether issues they might be having are because of some installed plugin or something in core